### PR TITLE
Fix Tricore disassembly performance

### PR DESCRIPTION
**Pull request by b1llow**
# SQUASH ME

**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

...

**Test plan**

...

**Closing issues**

...

### CODE REVIEW

**Comment by b1llow** [comment]
> I feel that disassembly of Tricore is significantly slower in Rizin and Cutter compared to other architectures. @rot127 could it be related to the tokenization? 


I tried to fix the problem you mentioned, could you test it. Or I can test it if you can share the binary file with me. @XVilka

**Comment by b1llow** [comment]
I found it does not require tokenize the disassembly for this issue https://github.com/rizinorg/rizin/issues/3664, so I removed the extra code.

Tested with this command, reduced from 1.71s to 1.25s

```
time rizin -escr.interactive=0 -Qc 'aaa;pd 10000 @ main' -A ~/Downloads/rizin_testcase/erika3app.elf > /dev/null
```

**Review by XVilka — COMMENTED** [review]
Yes, it's noticeably faster.
I got a new warning though during the analysis (can't share the file):
```
[0x80080020]> aaa
[x] Analyze all flags starting with sym. and entry0 (aa)
[WARNING: (../librz/analysis/p/../../asm/arch/tricore/tricore.inc:47):tricore_get_op_imm: code should not be reached
[x] Analyze function calls
[x] find and analyze function preludes
[WARNING: (../librz/analysis/p/../../asm/arch/tricore/tricore.inc:47):tricore_get_op_imm: code should not be reached
[x] Analyze len bytes of instructions for references
[x] Check for classes
```

**Comment by b1llow** [comment]
> Yes, it's noticeably faster. I got a new warning though during the analysis (can't share the file):
> 
> ```
> [0x80080020]> aaa
> [x] Analyze all flags starting with sym. and entry0 (aa)
> [WARNING: (../librz/analysis/p/../../asm/arch/tricore/tricore.inc:47):tricore_get_op_imm: code should not be reached
> [x] Analyze function calls
> [x] find and analyze function preludes
> [WARNING: (../librz/analysis/p/../../asm/arch/tricore/tricore.inc:47):tricore_get_op_imm: code should not be reached
> [x] Analyze len bytes of instructions for references
> [x] Check for classes
> ```

Could you rerun it and give the logs?

**Comment by XVilka** [comment]
> > Yes, it's noticeably faster. I got a new warning though during the analysis (can't share the file):
> > ```
> > [0x80080020]> aaa
> > [x] Analyze all flags starting with sym. and entry0 (aa)
> > [WARNING: (../librz/analysis/p/../../asm/arch/tricore/tricore.inc:47):tricore_get_op_imm: code should not be reached
> > [x] Analyze function calls
> > [x] find and analyze function preludes
> > [WARNING: (../librz/analysis/p/../../asm/arch/tricore/tricore.inc:47):tricore_get_op_imm: code should not be reached
> > [x] Analyze len bytes of instructions for references
> > [x] Check for classes
> > ```
> 
> Could you rerun it and give the logs?

```
[0x80080020]> aaa
[x] Analyze all flags starting with sym. and entry0 (aa)
[WARNING: failed to get imm operand: "sp, a0, a0"
[x] Analyze function calls
[x] find and analyze function preludes
[WARNING: failed to get imm operand: "sp, a0, a0"ces
[x] Analyze len bytes of instructions for references
[x] Check for classes
[x] Finding xrefs in noncode section with analysis.in=io.maps
[x] Analyze value pointers (aav)
```

**Comment by b1llow** [comment]
> > > Yes, it's noticeably faster. I got a new warning though during the analysis (can't share the file):
> > > ```
> > > [0x80080020]> aaa
> > > [x] Analyze all flags starting with sym. and entry0 (aa)
> > > [WARNING: (../librz/analysis/p/../../asm/arch/tricore/tricore.inc:47):tricore_get_op_imm: code should not be reached
> > > [x] Analyze function calls
> > > [x] find and analyze function preludes
> > > [WARNING: (../librz/analysis/p/../../asm/arch/tricore/tricore.inc:47):tricore_get_op_imm: code should not be reached
> > > [x] Analyze len bytes of instructions for references
> > > [x] Check for classes
> > > ```
> > 
> > 
> > Could you rerun it and give the logs?
> 
> ```
> [0x80080020]> aaa
> [x] Analyze all flags starting with sym. and entry0 (aa)
> [WARNING: failed to get imm operand: "sp, a0, a0"
> [x] Analyze function calls
> [x] find and analyze function preludes
> [WARNING: failed to get imm operand: "sp, a0, a0"ces
> [x] Analyze len bytes of instructions for references
> [x] Check for classes
> [x] Finding xrefs in noncode section with analysis.in=io.maps
> [x] Analyze value pointers (aav)
> ```

My bad, but there is still a problem with the log output. I need some log info for the new version.

**Comment by XVilka** [comment]
```
[0x80080020]> aaa
[x] Analyze all flags starting with sym. and entry0 (aa)
[WARNING: Failed to get operand1 [3]: "sub.a sp, a0, a0" [imm]
[x] Analyze function calls
[x] find and analyze function preludes
[WARNING: Failed to get operand1 [3]: "sub.a sp, a0, a0" [imm]
[x] Analyze len bytes of instructions for references
[x] Check for classes
[x] Finding xrefs in noncode section with analysis.in=io.maps
[x] Analyze value pointers (aav)
```

File: librz/asm/arch/tricore/tricore.inc [hunk 1]
```diff
@@ -44,12 +50,32 @@ static inline const char *tricore_get_op_regname(csh h, cs_insn *insn, int idx)
...
+			idx, tricore_op_count(insn), insn->mnemonic, insn->op_str);
 		return 0;
 	}
 	return op->imm;
 }
 
-static inline ut8 tricore_op_count(cs_insn *insn) {
-	return insn->detail->tricore.op_count;
+static inline csh tricore_setup_cs_handle(const char *cpu, const char *features) {
+	static csh handle = 0;
```
Comment (XVilka) [inline librz/asm/arch/tricore/tricore.inc:61]: Is it possible not to use `static`? Could be done in a separate PR. See https://github.com/rizinorg/rizin/issues/3721

**Review by XVilka — APPROVED** [review]
LGTM (tested and it works fine, no warnings, and also faster)

File: librz/asm/arch/tricore/tricore.inc (hunk 1 above)
Comment (b1llow) [reply librz/asm/arch/tricore/tricore.inc:61]: Possibly, but then we'd have to save the csh in RzAsm and RzAnalysis or choose not to cache the csh?

Comment (XVilka) [reply librz/asm/arch/tricore/tricore.inc:61]: Good point about caching. It's probably better to add a new member to the RzAnalysis/RzAsm plugin then. Like we have `il_config` for RzIL information, maybe we can have `arch_config` for different things, including capstone. @wargio @ret2libc @thestr4ng3r what do you think? See also https://github.com/rizinorg/rizin/commit/2c9a7008e9e30a10ff42c0422b1ef7d6579b8dab

Comment (XVilka) [reply librz/asm/arch/tricore/tricore.inc:61]: @imbillow nevermind, see `librz/analysis/p/analysis_arm_cs.c` for the example how to use context to save the handle.

**Review by XVilka — APPROVED** [review]

**Pull request merged**